### PR TITLE
Rush should create .tgz files instead of package.json for temp_modules

### PR DESCRIPTION
**Pull request by nickpape-msft**
Rush needs to make `file:` references to tgz files. NPM has decided to deprecate the behavior of the `file:./...` specified, which used to treat the folder as a project that needed to be installed. Instead, `file:./...` references now just create a symlink, **unless** one is referencing a `.tgz` file. PNPM has also decided to use this behavior.

We should merge this PR to:
1) maintain forwards compatibility with NPM
2) allow us to use PNPM

See:
https://github.com/npm/npm/pull/15900
and
https://github.com/pnpm/pnpm/issues/772

### CODE REVIEW

File: apps/rush/src/utilities/InstallManager.ts [hunk 1]
```diff
@@ -384,9 +382,79 @@ export default class InstallManager {
...
+
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2.new"
+      const tempProjectFolder: string = path.join(
+        path.dirname(rushProject.tempPackageTarballFilename),
+        unscopedTempProjectName + '.new');
+
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2\package.json"
+      const tempPackageJsonFilename: string = path.join(tempProjectFolder, RushConstants.packageJsonFilename);
+
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2.gzip"
```
Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: >gzip [](start = 63, length = 4)

*.tgz?

File: apps/rush/src/utilities/InstallManager.ts [hunk 2]
```diff
@@ -384,9 +382,79 @@ export default class InstallManager {
...
+
+      // we only want to overwrite the package if the existing tarball's package.json is different from tempPackageJson
+      let shouldOverwrite: boolean = true;
+      try {
+        // extract the tarball and compare the package.json directly
+        if (fsx.existsSync(tarballFile)) {
+
+          // ensure the folder we are about to extract into is clean
+          fsx.removeSync(extractedFolder);
+          fsx.mkdirpSync(extractedFolder);
```
Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: This fails intermittently due the virus scanner.

See Utilities.createFolderWithRetry() for more info

File: apps/rush/src/utilities/InstallManager.ts [hunk 3]
```diff
@@ -384,9 +382,79 @@ export default class InstallManager {
...
+          });
+
+          const extractedPackageJsonFilename: string =
+            path.join(extractedFolder, npmPackageFolder, RushConstants.packageJsonFilename);
+
+          // compare the extracted package.json with the one we are about to write
+          const oldBuffer: Buffer = fsx.readFileSync(extractedPackageJsonFilename);
+          const newBuffer: Buffer = new Buffer(JsonFile.normalize(tempPackageJson));
+
+          if (Buffer.compare(oldBuffer, newBuffer) === 0) {
```
Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: >(Buffer.compare(oldBuffer, newBuffer) === 0) [](start = 13, length = 44)

What about encodings?

File: apps/rush/src/utilities/InstallManager.ts [hunk 4]
```diff
@@ -384,9 +382,79 @@ export default class InstallManager {
...
+          }
+        }
+      } catch (error) {
+        // ignore the error, we will go ahead and create a new tarball
+      }
+
+      if (shouldOverwrite) {
+        // ensure the folder we are about to zip is clean
+        fsx.removeSync(tempProjectFolder);
+        fsx.mkdirpSync(tempProjectFolder);
```
Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: Same issue with virus scanner

File: apps/rush/src/utilities/InstallManager.ts [hunk 5]
```diff
@@ -384,9 +382,79 @@ export default class InstallManager {
...
+
+        // write the expected package.json file into the zip staging folder
+        JsonFile.saveJsonFile(tempPackageJson, tempPackageJsonFilename);
+
+        // create the new tarball, this overwrites the existing one
+        tar.create({
```
Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: >create [](start = 12, length = 6)

I would try once more to see if you can create the tarball without writing any files to disk.  Although the code will be a little more complicated, the runtime behavior will be faster and less likely to run into weird file locking issues.

Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: Also measure how long it takes to make 70 tarballs in sp-client

---
In reply to: [133607227](https://github.com/Microsoft/web-build-tools/pull/324#discussion_r133607227) [](ancestors = 133607227)

Comment (nickpape-msft) [inline apps/rush/src/utilities/InstallManager.ts]: Apparently the library I am using (which is the most-used tar library, written by none other than Isaac S. III) doesn't support in-memory creation. I will do a perf test in sp-client though.

---
In reply to: [133607247](https://github.com/Microsoft/web-build-tools/pull/324#discussion_r133607247) [](ancestors = 133607247,133607227)

Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: @nickpape-msft [archiver](https://www.npmjs.com/package/archiver) --> [tar-stream](https://www.npmjs.com/package/tar-stream) FYI

Comment (nickpape-msft) [inline apps/rush/src/utilities/InstallManager.ts]: Adds .33 seconds to rush install/generate time

---
In reply to: [133798619](https://github.com/Microsoft/web-build-tools/pull/324#discussion_r133798619) [](ancestors = 133798619,133607247,133607227)

File: apps/rush/src/utilities/InstallManager.ts (hunk 3 above)
Comment (nickpape-msft) [inline apps/rush/src/utilities/InstallManager.ts]: I borrowed this code from JsonFile, which you wrote. I suggest we fix this in all places

---
In reply to: [133606562](https://github.com/Microsoft/web-build-tools/pull/324#discussion_r133606562) [](ancestors = 133606562)

File: apps/rush/src/utilities/InstallManager.ts [hunk 6]
```diff
@@ -298,19 +302,16 @@ export default class InstallManager {
...
     );
+
     for (const rushProject of sortedRushProjects) {
       const packageJson: IPackageJson = rushProject.packageJson;
 
-      // Example: "C:\MyRepo\common\temp\projects\my-project-2"
-      const tempProjectFolder: string = path.dirname(rushProject.tempPackageJsonFilename);
-      fsx.mkdirsSync(tempProjectFolder);
+      // Example: "my-project-2"
+      const unscopedTempProjectName: string = path.basename(rushProject.tempPackageTarballFilename, '.tgz');
```
Comment (pgonzal) [inline apps/rush/src/utilities/InstallManager.ts]: >path.basename(rushProject.tempPackageTarballFilename, '.tgz'); [](start = 46, length = 62)

This is a weird design.  On line 393 you use "path.dirname(rushProject.tempPackageTarballFilename)" to calculate the same thing.

Why not just make "unscopedTempProjectName" your API property (e.g. called "tempProjectUniqueName"), and then construct the other derivatives from that.  I'd argue that these derivatives are all internal guts of InstallManager, not something that rush-lib consumers should be privy to.

installCommonModules() is a slightly tricky case, but you could just make a private helper function getTarballPath(project)

Comment (nickpape-msft) [inline apps/rush/src/utilities/InstallManager.ts]: I like this suggestion. Done.

**Pull request merged**